packages/ember-metal/lib/meta.js

```diff
@@ -0,0 +1,287 @@
+// Remove "use strict"; from transpiled module until
+// https://bugs.webkit.org/show_bug.cgi?id=138038 is fixed
+//
+'REMOVE_USE_STRICT: true';
+
+import isEnabled from 'ember-metal/features';
+import { protoMethods as listenerMethods } from 'ember-metal/meta_listeners';
+
+/**
+@module ember-metal
+*/
+
+/*
+ This declares several meta-programmed members on the Meta class. Such
+ meta!
+
+ In general, the `readable` variants will give you an object (if it
+ already exists) that you can read but should not modify. The
+ `writable` variants will give you a mutable object, and they will
+ create it if it didn't already exist.
+
+ The following methods will get generated metaprogrammatically, and
+ I'm including them here for greppability:
+
+ writableCache, readableCache, writableWatching, readableWatching,
+ peekWatching, clearWatching, writableMixins, readableMixins,
+ peekMixins, clearMixins, writableBindings, readableBindings,
+ peekBindings, clearBindings, writableValues, readableValues,
+ peekValues, clearValues, writableDeps, readableDeps, getAllDeps
+ writableChainWatchers, readableChainWatchers, writableChains,
+ readableChains
+
+*/
+let members = {
+  cache: ownMap,
+  watching: inheritedMap,
+  mixins: inheritedMap,
+  bindings: inheritedMap,
+  values: inheritedMap,
+  deps: inheritedMapOfMaps,
+  chainWatchers: ownCustomObject,
+  chains: inheritedCustomObject
+};
+
+let memberNames = Object.keys(members);
+
+function Meta(obj, parentMeta) {
+  // preallocate a slot for each member
+  for (let i = 0; i < memberNames.length; i++) {
+    this[memberProperty(memberNames[i])] = undefined;
```

Comment:
how about we move memberProperty to outside the constructor `memberProperties = Object.keys(members).map(memberProperty)`